### DISCUSSION TITLE
Best approach for building an admin interface to list / browse Identities

### DISCUSSION THREAD

**Discussion by devnet-io** [ory/kratos#1336] [Q&A] [answered]
I'm working on implementing administration sections and need to provide a mechanism for browsing identities. **Ideally the admin would be able to search, filter, and sort the list.** For the sake of simplicity assume this is a server side application. I can think of a three approaches to getting the necessary data to accomplish this (below). Are any of these preferred / recommended?

1. **Use hooks to maintain separate identity / user records**
- On registration a record would be inserted into a separate database. These records would be used to render the administration area, only retrieving data from kratos when necessary.
- For this option it would be very helpful if the Kratos identities endpoint could accept an array of ids to return. https://github.com/ory/kratos/blob/18518e93bad531fe57f0b073029aa8edd5822f60/identity/handler.go#L108 This would allow bulk fetching of traits in some situations.

2. **Consume the Kratos admin API directly**
- The app would call the /identities endpoint and use the returned data to render pages.
- For this approach i believe the Kratos api would need to support querying identities by traits, etc., perhaps similar to [Auth0's approach](https://auth0.com/docs/users/user-search/user-search-query-syntax).

3. **Write a custom service / api pointing to the Kratos database to serve identities**
- This approach is easy implement and allow for any parameters needed, but would have to be updated if the Kratos schema ever changes.

What would be recommended here?

**Comment by vinckr** [comment]
1 & 2 sound like good approaches in my opinion, but I am not a maintainer of this project, so proceed with caution 😉 .

We have an issue open that is related to this: https://github.com/ory/kratos/issues/655  Admin/Selfservice session management 

I think it is still some time until we can tackle that one (at the moment planned for 0.7,0.8), but I will keep my ears open if something is useful for this.

**Reply by devnet-io** [reply]
I would love to do 2 but i believe the current /identities endpoint would have to be enhanced. It might be possible get around this for 1 but I'd have to fetch traits for each row, essentially the n+1 problem.

**Comment by devnet-io** [comment, marked as answer]
I will open issues for the enhancements described in 1 and 2 if there are no objections.

**Reply by vinckr** [reply]
Definitely not! Appreciate that you are looking into this 👍

**Reply by vinckr** [reply]
Hey @devnet-io how has it been going, have not followed the progress sorry!
I will mark this as answered for now, but feel free to come back to the discussion any time.

Thanks a ton! 🐝

**Reply by eduhenke** [reply]
@devnet-io how have you solved this? care to share your solution? :grinning:

**Reply by hananint** [reply]
@devnet-io Will be great if you can share the solution.

**Reply by vinckr** [reply]
Hey @eduhenke @hananint 
Also, check out these issues, they are a bit further along:
https://github.com/ory/kratos/issues/249
https://github.com/ory/kratos/issues/655

Always good to hear your input if you have any thoughts on the proposals so far!
